activemodel/lib/active_model/errors.rb

```diff
@@ -509,46 +585,116 @@ def generate_message(attribute, type = :invalid, options = {})
-    end
-
     def marshal_load(array) # :nodoc:
-      @base, @messages, @details = array
-      apply_default_array(@messages)
-      apply_default_array(@details)
+      # Rails 5
+      @errors = []
+      @base = array[0]
+      add_from_legacy_details_hash(array[2])
     end
 
     def init_with(coder) # :nodoc:
-      coder.map.each { |k, v| instance_variable_set(:"@#{k}", v) }
-      @details ||= {}
-      apply_default_array(@messages)
-      apply_default_array(@details)
+      data = coder.map
+
+      data.each { |k, v|
+        next if LEGACY_ATTRIBUTES.include?(k.to_sym)
+        instance_variable_set(:"@#{k}", v)
+      }
+
+      @errors ||= []
+
+      # Legacy support Rails 5.x details hash
+      add_from_legacy_details_hash(data["details"]) if data.key?("details")
     end
 
-  private
-    def normalize_message(attribute, message, options)
-      case message
-      when Symbol
-        generate_message(attribute, message, options.except(*CALLBACKS_OPTIONS))
-      else
-        message
+    private
+
+      def normalize_detail(type, options)
+        { error: type }.merge(options.except(*MESSAGE_OPTIONS))
+      end
+
+      # Error type can appear as <tt>type</tt> or <tt>options[:message]</tt>.
+      # Message or type can also be dynamic.
+      # This method evaluates them and normalize type/message to the appropriate place.
+      def normalize_arguments(attribute, type, **options)
+        options = options.except(*CALLBACKS_OPTIONS)
+
+        # Evaluate proc first
+        if type.respond_to?(:call)
+          type = type.call(@base, options)
+        end
+
+        if type && !type.is_a?(Symbol)
+          options[:message] ||= type
+          type = nil
+        end
+
+        [attribute.to_sym, type, options]
+      end
+
+      def add_from_legacy_details_hash(details)
+        details.each { |attribute, errors|
+          errors.each { |error|
+            type = error.delete(:error)
+            add(attribute, type, error)
+          }
+        }
       end
+
+      def deprecation_removal_warning(method_name)
+        ActiveSupport::Deprecation.warn("ActiveModel::Errors##{method_name} is deprecated and will be removed in Rails 6.1")
+      end
+
+      def deprecation_rename_warning(old_method_name, new_method_name)
+        ActiveSupport::Deprecation.warn("ActiveModel::Errors##{old_method_name} is deprecated. Please call ##{new_method_name} instead.")
+      end
+  end
+
+  class DeprecationHandlingMessageHash < SimpleDelegator
+    def initialize(errors)
+      @errors = errors
+      super(prepare_content)
     end
 
-    def normalize_detail(message, options)
-      { error: message }.merge(options.except(*CALLBACKS_OPTIONS + MESSAGE_OPTIONS))
+    def []=(attribute, value)
+      ActiveSupport::Deprecation.warn("Calling `[]=` to an ActiveModel::Errors is deprecated. Please call `ActiveModel::Errors#add` instead.")
+
+      Array(value).each do |message|
+        @errors.add(attribute, message)
+      end
```

Comment:
If there are existing errors on `attribute`, this will just append to them. This feels like a behaviour change?

Maybe we should have `@errors.add(attribute)` before this loop?

Reply:
I agree, but there exists this test:

```ruby
  test "values returns an array of messages" do
    errors = ActiveModel::Errors.new(self)
    errors.messages[:foo] = "omg"
    errors.messages[:baz] = "zomg"

    assert_equal ["omg", "zomg"], errors.values
  end
```

Weird, I know O_O

Reply:
So this is a little complicated. The `errors[]=` was deprecated in 5.0, and removed in 5.1.

In 5.0, it actually `appended` the message: https://api.rubyonrails.org/v5.0.0/classes/ActiveModel/Errors.html#method-i-5B-5D-3D

Now in 5.1, the method was removed from API, but user can bypass them by calling `errors.messages[]=` instead. It overrides the `messages` array as you mentioned. This is actually a bad thing, for example:

```ruby
errors.messages[:foo] = "zomg"
assert_equal ["zomg"], errors[:foo] # This fails because it is not wrapped in an array
```

`errors[:foo]` now does not return an array, which can break errors rendering code.

So now I feel actually appending is a better solution, especially when we deliberately removed the `[]=` API. WDYT?